cmd/contact-auditor/main.go

```diff
@@ -10,101 +12,147 @@ import (
-	    WHERE r.contact != '[]'
-		    AND c.expires >= :expireCutoff`,
-		map[string]interface{}{
-			"expireCutoff": c.clk.Now().Add(-c.grace),
-		})
-	if err != nil {
-		return nil, fmt.Errorf("error while querying database: %s", err)
+	for _, contact := range r.contacts {
+		if strings.HasPrefix(contact, "mailto:") {
+			err := policy.ValidEmail(strings.TrimPrefix(contact, "mailto:"))
+			if err != nil {
+				writeProb(contact, err.Error())
+			}
+		} else {
+			writeProb(contact, "missing 'mailto:' prefix")
+		}
+	}
+
+	if probsBuff.String() != "" {
+		return errors.New(probsBuff.String())
 	}
-	return holder, nil
+	return nil
 }
 
-// unmarshalAddresses unmarshalls the `Contact` field of the inner
-// `queryResult` and extracts the email addresses.
-func (r *queryResult) unmarshalAddresses() error {
-	var contactFields []string
-	err := json.Unmarshal(r.Contact, &contactFields)
-	if err != nil {
-		return err
+type results struct {
+	entries []*result
+	logger  blog.Logger
+}
+
+func (r *results) unmarshalContacts() {
+	for _, result := range r.entries {
+		err := result.unmarshalContact()
+		if err != nil {
+			r.logger.Errf("Unmarshal failed for ID: %d due to: %s", result.ID, err)
+			continue
+		}
 	}
-	for _, entry := range contactFields {
-		if strings.HasPrefix(entry, "mailto:") {
-			r.addresses = append(r.addresses, strings.TrimPrefix(entry, "mailto:"))
+}
+
+func (r results) validateContacts() {
+	for _, result := range r.entries {
+		err := result.validateContact()
+		if err != nil {
+			r.logger.Errf("Validation failed for ID: %s due to: %s", result.ID, err)
+			continue
 		}
 	}
-	return nil
 }
 
-// run extracts email addresses from the database and attempts to
-// validate each.
-func (c contactAuditor) run() (queryResults, error) {
-	results, err := c.collectContacts()
+// collectContacts queries the database for all IDs and contacts with
+// unexpired certificates.
+func (c contactAuditor) collectContacts() (*results, error) {
+	// Setting isolation level to `READ UNCOMMITTED` improves
+	// performance at the cost of consistency.
+	_, err := c.db.Exec("SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;")
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error setting db trancaction isolation level: %s", err)
+	}
+
+	rows, err := c.db.Query(
+		fmt.Sprintf(
+			"%s '%s';",
+			`SELECT DISTINCT r.id, r.contact
+			FROM registrations AS r
+				INNER JOIN certificates AS c on c.registrationID = r.id
+			WHERE r.contact NOT IN ('[]', 'null')
+				AND c.expires >=`,
+			c.clk.Now().Add(-c.grace).Format("2006-01-02T15:04:05Z")),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error performing db query: %s", err)
 	}
-	for _, result := range results {
-		err = result.unmarshalAddresses()
+
+	c.logger.Infof("Gathering query results")
+	results := results{logger: c.logger}
+	for rows.Next() {
+		var result result
+		err := rows.Scan(&result.ID, &result.Contact)
```

Comment:
We're being careful to now let the network buffer below by getting underneath the ORM, but this is still now collecting all the results in memory before doing work.

Consider instead, in this loop, calling `result.unmarshalContact()` and `result.validateContact()` and do not collect the results into a `results` object. In fact, you can then remove that object. Since validation results are put into the log, you do not need to retain the `result` objects anywhere for prod.

However, since you use them for tests, I would recommend that you put each result into a channel. For production, you can spin a go thread that just takes each result and drops it on the floor. In tests, you can evaluate the output of the channel when necessary.

Reply:
Ah 👍🏻